src/network/resolver.rs

```diff
@@ -0,0 +1,89 @@
+use std::{
+    collections::HashMap,
+    net::Ipv4Addr,
+    sync::{Arc, Mutex},
+    thread::{Builder, JoinHandle},
+};
+use tokio::{
+    runtime::Runtime,
+    sync::mpsc::{self, Sender},
+};
+use trust_dns_resolver::AsyncResolver;
+
+pub type Result<T> = std::result::Result<T, failure::Error>;
+pub type IpTable = Arc<Mutex<HashMap<Ipv4Addr, String>>>;
+
+const CHANNEL_SIZE: usize = 1_000;
+
+/// Lookup translates between IP addresses and hostnames asynchronously.
+pub trait Lookup: Send {
+    /// Lookup the hostname for the given IP address asynchronously.
+    fn lookup(&mut self, addr: Ipv4Addr);
+}
+
+/// An asynchronous implementation of the Lookup trait.
+pub struct Resolver {
+    tx: Option<Sender<Ipv4Addr>>,
+    handle: Option<JoinHandle<()>>,
+}
+
+impl Resolver {
+    /// Create a new resolver with the given address translation table.
+    pub fn new(table: IpTable) -> Result<Self> {
+        let runtime = Runtime::new()?;
+        let (tx, mut rx) = mpsc::channel::<Ipv4Addr>(CHANNEL_SIZE);
+        let (resolver, background) = AsyncResolver::from_system_conf()?;
+
+        let handle = Builder::new().name("resolver".into()).spawn({
+            move || {
+                runtime.block_on(async {
+                    let resolver = Arc::new(resolver);
+                    runtime.spawn(background);
+
+                    while let Some(ip) = rx.recv().await {
```

Comment:
Tokio (and async/await in general) are new to me - so maybe I don't understand this right - please correct me if I'm wrong. :)
It seems to me that the requests here still happen consecutively, that this blocks until we receive an IP to resolve, then we resolve it, and only then do we request another, right?

Reply:
It's new stuff for me too :) But if I'm not mistaken, the workflow goes like this:
1) The stream will block until it receives an IP, as you said.
2) For each IP received we will spawn a new task that will await the resolution, but this will happen concurrently.

So if we receive e.g. 4 IPs we will create 4 tasks that will run in the thread pool. They will each send a request to the DNS server until they have to block for the response, at this point each task will return `Poll:Pending` and will be suspended. Tokio will "register interest" in the responses using e.g. `epoll` in Linux. At some point in the future a response for one of them will be received, the tokio reactor will be notified and will wake the corresponding task, and then the insertion in the map will take place. This will be repeated for the other 3 tasks in the example as the responses arrive.

Related to this part of the code, I was thinking about restoring the interface of the dns queue that was taking a vector of IPs instead of taking them one by one; this should reduce unnecessary contention on the channel.